ipld/amt/src/diff.rs

```diff
@@ -0,0 +1,387 @@
+// Copyright 2021-2023 Protocol Labs
+// Copyright 2019-2022 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use anyhow::Context;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::CborStore;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::node::CollapsedNode;
+
+use super::*;
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ChangeType {
+    Add,
+    Remove,
+    Modify,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Change<V> {
+    pub change_type: ChangeType,
+    pub key: u64,
+    pub before: Option<V>,
+    pub after: Option<V>,
+}
+
+struct NodeContext<'bs, BS> {
+    pub height: u32,
+    pub bit_width: u32,
+    pub store: &'bs BS,
+}
+
+impl<'bs, BS> NodeContext<'bs, BS> {
+    fn nodes_at_height(&self) -> u64 {
+        nodes_for_height(self.bit_width, self.height)
+    }
+}
+
+impl<'bs, V, BS> From<&'bs Amt<V, BS>> for NodeContext<'bs, BS> {
+    fn from(value: &'bs Amt<V, BS>) -> Self {
+        Self {
+            height: value.height(),
+            bit_width: value.bit_width(),
+            store: &value.block_store,
+        }
+    }
+}
+
+/// Returns a set of changes that transform node 'a' into node 'b'.
+/// Ported from <https://github.com/filecoin-project/go-amt-ipld/blob/master/diff.go#L41>
+pub fn diff<V, BS>(prev_amt: &Amt<V, BS>, curr_amt: &Amt<V, BS>) -> anyhow::Result<Vec<Change<V>>>
```

Comment:
nit: I'd consider allowing these to have different blockstore types (if possible).

Reply:
Possibly allowing them to have different value types as long as their value types implement `Eq`.

Reply:
Although that second part is probably overkill.

Reply:
Done. Regarding `Eq` derives, the Go implementation use cid for eq check

```rust
(Some(prev_link), Some(curr_link)) => {
                            let prev_cid = match prev_link {
                                node::Link::Cid { cid, .. } => cid,
                                _ => {
                                    anyhow::bail!("Unchanged link expected")
                                }
                            };
                            let curr_cid = match curr_link {
                                node::Link::Cid { cid, .. } => cid,
                                _ => {
                                    anyhow::bail!("Unchanged link expected")
                                }
                            };

                            if prev_cid == curr_cid {
                                continue;
                            }
                             ...
}
```